# Data Cleaning
This notebook will focus on the preprocessing of the datasets before using them in the two recommendation systems (Content-based and collaborative filtering) I will be developing later on.

In [52]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Suppress scientific notation of pandas
pd.set_option('display.float_format', lambda x: '%.0f' % x)

## Anime Dataset

Keys:
- anime_id - myanimelist.net's unique id identifying an anime.
- name - full name of anime.
- genre - comma separated list of genres for this anime.
- type - movie, TV, OVA, etc.
- episodes - how many episodes in this show. (1 if movie).
- rating - average rating out of 10 for this anime.
- members - number of community members that are in this anime's
"group".

In [53]:
# load anime dataset
anime_df = pd.read_csv("datasets/anime.csv")

anime_df.head()

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1       9   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64       9   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51       9   
3                                   Sci-Fi, Thriller     TV       24       9   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51       9   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266

In [55]:
# number of rows and columns in the dataframe
anime_df.shape

(12294, 7)

In [56]:
# are the columns using suitable datatypes
anime_df.dtypes

anime_id      int64
name         object
genre        object
type         object
episodes     object
rating      float64
members       int64
dtype: object

In [57]:
anime_df[["rating", "members"]].describe()

rating  members
count   12064    12294
mean        6    18071
std         1    54821
min         2        5
25%         6      225
50%         7     1550
75%         7     9437
max        10  1013917

In [58]:
# Check which rows have missing values
anime_df.isnull().any()

anime_id    False
name        False
genre        True
type         True
episodes    False
rating       True
members     False
dtype: bool

In [59]:
# How many missing values do we have for each column?
anime_df.isnull().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

It seems like there are missing values for the genre(62), type(25), and rating(230) columns. For a recommendation system, missing values may make a content based filtering system less inaccurate as features like the genre a film may influence the enjoyment one may have viewing a certain anime. 

<br>
Though this may be the case, I believe that for now, the anime dataset does not need any more cleaning.




## Ratings Dataset

Keys:
- user_id - non identifiable randomly generated user id.
- anime_id - the anime that this user has rated.
- rating - rating out of 10 this user has assigned (-1 if the user watched it but didn't assign a rating).

In [60]:
# load ratings dataset
rating_df = pd.read_csv("datasets/rating.csv")

rating_df.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [61]:
rating_df.tail()

user_id  anime_id  rating
7813732    73515     16512       7
7813733    73515     17187       9
7813734    73515     22145      10
7813735    73516       790       9
7813736    73516      8074       9

In [62]:
rating_df.shape

(7813737, 3)

In [67]:
rating_df.dtypes

user_id       int64
anime_id      int64
rating      float64
dtype: object

In [64]:
rating_df.isnull().any()

user_id     False
anime_id    False
rating      False
dtype: bool

Using -1 as the value for a no-rating may skew future analysis and the building of the recommender. Instead f using -1, I will replace all ratings of -1 with a null value.

In [90]:
rating_df["rating"].replace({-1: np.nan}, inplace=True)
values = rating_df["rating"].unique()
values.sort
print(values)

[nan 10.  8.  6.  9.  7.  3.  5.  4.  1.  2.]


Now the ratings has values from 1-10 and nan for empty ratings.

The ratings dataset doesn't seem to need any more cleaning unless other issues arise.